src/Symfony/Component/HttpClient/ServerSentEvents/EventSource.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+namespace Symfony\Component\HttpClient\ServerSentEvents;
+
+use Symfony\Component\HttpClient\HttpClient;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+final class EventSource implements EventSourceInterface
+{
+    private $url;
+    private $client;
+    private $reconnectionTime = 5;
+    private $shouldClose = false;
+    private $lastEventId = null;
+    private $response = null;
+
+    const CONNECTING = 0;
+    const OPEN = 1;
+    const CLOSED = 2;
+
+    public $readyState = self::CLOSED;
+
+    public function __construct(string $url, array $defaultOptions = [], HttpClientInterface $client = null)
+    {
+        $this->url = $url;
+        $this->client = $client ?: HttpClient::create();
+    }
+
+    public function getHttpClient(): HttpClientInterface
+    {
+        return $this->client;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getMessages(): \Iterator
+    {
+        while (false === $this->shouldClose) {
+            $response = $this->request();
+            $buffer = '';
+            foreach ($this->client->stream($response, $this->reconnectionTime) as $chunk) {
+                if ($chunk->isTimeout()) {
+                    break;
+                }
+
+                // We have to gather our own Buffer that may be longer then a chunk
+                $buffer .= $chunk->getContent();
+
+                if (!$buffer) {
+                    // var_dump('closed');
+                    // connection closed, should we throw ?
+                    return null;
+                }
+
+                if (self::OPEN !== $this->readyState) {
+                    $this->readyState = self::OPEN;
+                }
+
+                // If the line starts with a U+003A COLON character (':') ignore the line
+                if (':' === $buffer[0]) {
+                    $buffer = '';
+                    continue;
+                }
+
+                // Process if we got new line ending
+                $end = substr($buffer, -1);
+                if ("\n" !== $end && "\r" !== $end) {
+                    continue;
+                }
+
+                $message = MessageEvent::parse($buffer);
+                if (null !== $message->getId()) {
+                    $this->requestArgs[2]['headers']['Last-Event-ID'] = $message->getId();
+                }
+
+                $buffer = '';
+
+                if (null !== $message->getRetry()) {
+                    $this->reconnectionTime = $message->retry;
+                }
+
+                yield $message;
+            }
+        }
+    }
+
+    public function close()
```

Comment:
This partially answers my question on the constructor: this class is mutable, so it cannot be a service. Don't we want to make this a possible service instead?

Reply:
@dunglas was thinking about adding some factory:

```
class EventSourceFactory {
   construct($defaultClient)
   createEventSource($url)
}
```

Indeed, having the URL in the constructor of this class is closer to the specification but I'm all ears for suggestions.

Reply:
My preference would be a design that extends HttpClientInterface, by adding a single `connect()` method (or similar name). This method would ensure that all needed headers are sent, and would return a ResponseInterface that wraps the original response from the client. The stream() method would then ensure to yield chunks that match message boundaries when yielding these wrapped responses. It could also stream regular responses of course.

This is more complex to implement but it would allow monitoring both SSE and regular responses at once.